README.md
# Dart Kafka

[![Build Status](https://travis-ci.org/pulyaevskiy/kafkabr.svg?branch=master)](https://travis-ci.org/pulyaevskiy/kafkabr)
[![Coverage](https://codecov.io/gh/pulyaevskiy/kafkabr/branch/master/graph/badge.svg)](https://codecov.io/gh/pulyaevskiy/kafkabr)
[![License](https://img.shields.io/badge/license-BSD--2-blue.svg)](https://raw.githubusercontent.com/pulyaevskiy/kafkabr/master/LICENSE)

Kafka client library written in Dart.

### Current status

This library is a work-in-progress.
Currently all the updates are happening in `kafka-0.10` branch:

* Support for Kafka 0.10 APIs including Group Membership API
* Implementation of `HighLevelConsumer` capable of automatic load-balancing
  and re-distribution of topics/partitions in case of failures.
* Better testing framework.
* Isolate-based distribution of consumer group members for better utilization
  of system resources.

Master branch currently targets 0.8.x versions of Kafka server.

### Things that are not supported yet.

* Snappy compression.

## Installation

There is no Pub package yet, but it will be published as soon as APIs are
stable enough.

For now you can use git dependency in your `pubspec.yaml`:

```yaml
dependencies:
  kafka:
    git: https://github.com/armando-couto/kafkabr.git
```

And then import it as usual:

```dart
import 'package:kafkabr/kafka.dart';
```

## Features

This library provides several high-level API objects to interact with Kafka:

* __KafkaSession__ - responsible for managing connections to Kafka brokers and
  coordinating all requests. Also provides access to metadata information.
* __Producer__ - publishes messages to Kafka topics
* __Consumer__ - consumes messages from Kafka topics and stores it's state (current
  offsets). Leverages ConsumerMetadata API via ConsumerGroup.
* __Fetcher__ - consumes messages from Kafka without storing state.
* __OffsetMaster__ - provides convenience on top of Offset API allowing to easily
  retrieve earliest and latest offsets of particular topic-partitions.
* __ConsumerGroup__ - provides convenience on top of Consumer Metadata API to easily
  fetch or commit consumer offsets.

## Producer

Simple implementation of Kafka producer. Supports auto-detection of leaders for
topic-partitions and creates separate `ProduceRequest`s for each broker.
Requests are sent in parallel and all responses are aggregated in special
`ProduceResult` object.

```dart
// file:produce.dart
import 'dart:io';
import 'package:kafkabr/kafka.dart';

main(List<String> arguments) async {
  var host = new ContactPoint('127.0.0.1', 9092);
  var session = new KafkaSession([host]);

  var producer = new Producer(session, 1, 1000);
  var result = await producer.produce([
    new ProduceEnvelope('topicName', 0, [new Message('msgForPartition0'.codeUnits)]),
    new ProduceEnvelope('topicName', 1, [new Message('msgForPartition1'.codeUnits)])
  ]);
  print(result.hasErrors);
  print(result.offsets);
  session.close(); // make sure to always close the session when the work is done.
}
```

Result:

```shell
$ dart produce.dart
$ false
$ {dartKafkaTest: {0: 213075, 1: 201680}}
```

## Consumer

High-level implementation of Kafka consumer which stores it's state using
Kafka's ConsumerMetadata API.

> If you don't want to keep state of consumed offsets take a look at `Fetcher`
> which was designed specifically for this use case.

Consumer returns messages as a `Stream`, so all standard stream operations
should be applicable. However Kafka topics are ordered streams of messages
with sequential offsets. Consumer implementation allows to preserve order of
messages received from server. For this purpose all messages are wrapped in
special `MessageEnvelope` object with following methods:

```
/// Signals to consumer that message has been processed and it's offset can
/// be committed.
void commit(String metadata);

/// Signals that message has been processed and we are ready for
/// the next one. Offset of this message will **not** be committed.
void ack();

/// Signals to consumer to cancel any further deliveries and close the stream.
void cancel();
```

One must call `commit()` or `ack()` for each processed message, otherwise
Consumer won't send the next message to the stream.

Simplest example of a consumer:

```dart
import 'dart:io';
import 'dart:async';
import 'package:kafkabr/kafka.dart';

void main(List<String> arguments) async {
  var host = new ContactPoint('127.0.0.1', 9092);
  var session = new KafkaSession([host]);
  var group = new ConsumerGroup(session, 'consumerGroupName');
  var topics = {
    'topicName': [0, 1] // list of partitions to consume from.
  };

  var consumer = new Consumer(session, group, topics, 100, 1);
  await for (MessageEnvelope envelope in consumer.consume(limit: 3)) {
    // Assuming that messages were produces by Producer from previous example.
    var value = new String.fromCharCodes(envelope.message.value);
    print('Got message: ${envelope.offset}, ${value}');
    envelope.commit('metadata'); // Important.
  }
  session.close(); // make sure to always close the session when the work is done.
}
```

It is also possible to consume messages in batches for improved efficiency:

```dart
import 'dart:io';
import 'dart:async';
import 'package:kafkabr/kafka.dart';

void main(List<String> arguments) async {
  var host = new ContactPoint('127.0.0.1', 9092);
  var session = new KafkaSession([host]);
  var group = new ConsumerGroup(session, 'consumerGroupName');
  var topics = {
    'topicName': [0, 1] // list of partitions to consume from.
  };

  var consumer = new Consumer(session, group, topics, 100, 1);
  await for (BatchEnvelope batch in consumer.batchConsume(20)) {
    batch.items.forEach((MessageEnvelope envelope) {
      // use envelope as usual
    });
    batch.commit('metadata'); // use batch control methods instead of individual messages.
  }
  session.close(); // make sure to always close the session when the work is done.
}
```

### Consumer offset reset strategy

Due to the fact that Kafka topics can be configured to delete old messages
periodically, it is possible that your consumer offset may become invalid (
just because there is no such message/offset in Kafka topic anymore).

In such cases `Consumer` provides configurable strategy with following options:

* `OffsetOutOfRangeBehavior.throwError`
* `OffsetOutOfRangeBehavior.resetToEarliest` (default)
* `OffsetOutOfRangeBehavior.resetToLatest`

By default if it gets `OffsetOutOfRange` server error it will reset it's offsets
to earliest available in the consumed topic and partitions, which essentially
means consuming all available messages from the beginning.

To modify this behavior simply set `onOffsetOutOfRange` property of consumer to
one of the above values:

```
var consumer = new Consumer(session, group, topics, 100, 1);
consumer.onOffsetOutOfRange = OffsetOutOfRangeBehavior.throwError;
```

## Supported protocol versions

Current version targets version `1.1.1` of the Kafka protocol. There is no plans
to support earlier versions.

## License

BSD-2

### Publish
```shell
dart pub publish --dry-run
dart pub publish
```

CHANGELOG.md
## 1.1.1

* Comunicação de um kafka remotamente.

lib/src/protocol/offset_commit_api.dart
part of kafka.protocol;

/// Kafka OffsetCommitRequest.
class OffsetCommitRequest extends KafkaRequest {
  /// API key of [OffsetCommitRequest].
  final int apiKey = 8;

  /// API version of [OffsetCommitRequest].
  final int apiVersion = 1;

  /// Name of the consumer group.
  final String consumerGroup;

  /// Generation ID of the consumer group.
  final int consumerGroupGenerationId;

  /// ID of the consumer.
  final String consumerId;

  /// Time period in msec to retain the offset.
  final int retentionTime;

  /// List of consumer offsets to be committed.
  final List<ConsumerOffset> offsets;

  /// Creates new instance of [OffsetCommitRequest].
  ///
  /// [host] must be current coordinator broker for [consumerGroup].
  OffsetCommitRequest(this.consumerGroup, this.offsets, this.consumerGroupGenerationId, this.consumerId, this.retentionTime) : super();

  @override
  List<int> toBytes() {
    var builder = new KafkaBytesBuilder.withRequestHeader(apiKey, apiVersion, correlationId);

    // TODO: replace groupBy with ListMultimap
    // ignore: STRONG_MODE_DOWN_CAST_COMPOSITE
    Map<String, List<ConsumerOffset>> groupedByTopic = groupBy(offsets, (o) => o.topicName);
    var timestamp = DateTime.now().millisecondsSinceEpoch;
    builder.addString(consumerGroup);
    builder.addInt32(consumerGroupGenerationId);
    builder.addString(consumerId);
    builder.addInt32(groupedByTopic.length);
    groupedByTopic.forEach((topicName, partitionOffsets) {
      builder.addString(topicName);
      builder.addInt32(partitionOffsets.length);
      partitionOffsets.forEach((p) {
        builder.addInt32(p.partitionId);
        builder.addInt64(p.offset);
        builder.addInt64(timestamp);
        builder.addString(p.metadata);
      });
    });

    var body = builder.takeBytes();
    builder.addBytes(body);

    return builder.takeBytes();
  }

  @override
  createResponse(List<int> data) {
    return new OffsetCommitResponse.fromData(data);
  }
}

/// Kafka OffsetCommitResponse.
class OffsetCommitResponse {
  final List<OffsetCommitResult> offsets;

  OffsetCommitResponse._(this.offsets);

  factory OffsetCommitResponse.fromData(List<int> data) {
    List<OffsetCommitResult> offsets = [];
    var reader = new KafkaBytesReader.fromBytes(data);
    var size = reader.readInt32();
    assert(size == data.length - 4);

    reader.readInt32(); // correlationId
    var count = reader.readInt32();
    while (count > 0) {
      var topicName = reader.readString();
      var partitionCount = reader.readInt32();
      while (partitionCount > 0) {
        var partitionId = reader.readInt32();
        var errorCode = reader.readInt16();
        offsets.add(new OffsetCommitResult(topicName, partitionId, errorCode));
        partitionCount--;
      }
      count--;
    }

    return new OffsetCommitResponse._(offsets);
  }
}

/// Data structure representing result of commiting of consumer offset.
class OffsetCommitResult {
  final String topicName;
  final int partitionId;
  final int errorCode;

  OffsetCommitResult(this.topicName, this.partitionId, this.errorCode);
}


exemplo/lib/main.dart
import 'dart:ffi';

import 'package:flutter/material.dart';
import 'package:kafkabr/kafka.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  Future<void> _incrementCounter() async {
    setState(() async {
      // This call to setState tells the Flutter framework that something has
      // changed in this State, which causes it to rerun the build method below
      // so that the display can reflect the updated values. If we changed
      // _counter without calling setState(), then the build method would not be
      // called again, and so nothing would appear to happen.
      _counter++;

      var host = new ContactPoint('192.168.0.111', 9092);
      var session = new KafkaSession([host]);

      var producer = new Producer(session, 1, 1000);
      var result = await producer.produce([
        // new ProduceEnvelope('topicName', 0, [new Message('msgForPartition0'.codeUnits)]),
        new ProduceEnvelope('topicName', 0, [new Message('{"name": "Armando uepa"}'.codeUnits, key: 'body'.codeUnits)])
      ]);
      print(result.hasErrors);
      print(result.offsets);


      var group = new ConsumerGroup(session, 'consumerGroupName');
      var topics = {
        'topicName': <int>{0} // list of partitions to consume from.
      };

      var consumer = new Consumer(session, group, topics, 100, 1);
      await for (BatchEnvelope batch in consumer.batchConsume(20)) {
        batch.items.forEach((MessageEnvelope envelope) {
          print(envelope.message);
        });
        batch.commit('metadata'); // use batch control methods instead of individual messages.
      }
      session.close(); // make sure to always close the session when the work is done.

      session.close(); // make sure to always close the session when the work is done.
    });
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Center(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: Column(
          // Column is also a layout widget. It takes a list of children and
          // arranges them vertically. By default, it sizes itself to fit its
          // children horizontally, and tries to be as tall as its parent.
          //
          // Invoke "debug painting" (press "p" in the console, choose the
          // "Toggle Debug Paint" action from the Flutter Inspector in Android
          // Studio, or the "Toggle Debug Paint" command in Visual Studio Code)
          // to see the wireframe for each widget.
          //
          // Column has various properties to control how it sizes itself and
          // how it positions its children. Here we use mainAxisAlignment to
          // center the children vertically; the main axis here is the vertical
          // axis because Columns are vertical (the cross axis would be
          // horizontal).
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            const Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '$_counter',
              style: Theme
                  .of(context)
                  .textTheme
                  .headlineMedium,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: const Icon(Icons.add),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
